A49-dynamic-connection-scaling.md

```diff
@@ -0,0 +1,211 @@
+connection is modified to not queue RPCs, but report the scenario back to the
+subchannel instead, or to coordinate the `SETTINGS` frame ACK with the
+subchannel.  Such changes are out of scope for this design, but may be
+considered in the future.  Until this is designed and implemented, it is
+acceptable to queue RPCs on a connection due to this race, which is expected to
+be rare.
+
+If no connection is available for an RPC, the RPC must be queued in the
+subchannel until a connection is available for it.  This queue must be roughly
+fair: RPCs must be dispatched in the order in which they are received into the
+queue, acknowledging that timing between threads may lead to concurrent RPCs
+being added to the queue in an arbitrary order.  See "Potential Future Work" for
+a possible adjustment to this queuing strategy.
+
+### Settings
+
+#### Service Config Settings
+
+The top-level service config of gRPC will be modified for this feature as follows:
+
+```
+{
+	// ...
+	// existing fields, e.g. "methodConfig" and "loadBalancingConfig"
+	// ...
+
+	// Settings to control dynamic connection scaling.  For more details,
+    // please refer to gRFC AXX (this one; XX = TBD).
+	"connectionScaling": {
+		// Maximum connections gRPC will maintain for each subchannel in
+        // this channel.  When no streams are available for an RPC in a
+        // subchannel, gRPC will automatically create new connections up
+        // to this limit.  If this value changes during the life of a
+        // channel, existing subchannels will be updated to reflect
+        // the change.  No connections will be closed as a result of
+        // lowering this value.  Down-scaling will only happen as
+        // connections are lost naturally.
+        //
+        // Must be a whole number greater than 1.  Values higher than
+        // the configured limit (by default, 10) will be clamped.
+		"maxConnectionsPerSubchannel": number
+	}
+}
+```
+
+#### Channel Settings
+
+The `maxConnectionsPerSubchannel` setting from service config has a natural
+limit of 10.  A channel setting should be provided to allow a client application
+to raise or lower this limit.
+
+### Temporary environment variable protection
+
+Until sufficient testing is performed, this feature will initially be disabled
+by default, and only enabled if `GRPC_EXPERIMENTAL_CONNECTION_SCALING=true`.
+
+### Subchannel Sharing
+
+Subchannels may be shared across channels.  The `maxConnectionsPerSubchannel`
+setting will be applied to a shared subchannel and that subchannel will use the
+highest maximum of all channels currently sharing it.  If channels sharing the
+subchannel are closed, then the remaining channels' settings will be combined to
+determine the new max.  If the max is lowered to less than the number of open
+connections, the excess connections will not be closed as a result.
+Down-scaling will only happen as connections are lost naturally.
```

Comment:
I think this part can be simplified. In Node, and I think also Core, subchannels are only shared among channels with the same target and channel arguments. Two such channels should never have different service configs except in brief windows when updates may propagate to separate resolvers at slightly different times. So, there probably isn't much value in tracking which setting value is associated with which channel, and instead we can just use the latest value the subchannel has seen from any channel.